inc/Engine/CriticalPath/APIClient.php

```diff
@@ -0,0 +1,230 @@
+<?php
+namespace WP_Rocket\Engine\CriticalPath;
+
+use WP_Error;
+
+class APIClient {
+
+	/**
+	 * Constant url for Critical Path API job.
+	 */
+	const API_URL = 'https://cpcss.wp-rocket.me/api/job/';
+
+	/**
+	 * Sends a generation request to the Critical Path API
+	 *
+	 * @since 3.6
+	 *
+	 * @param string $url The URL to send a CPCSS generation request for.
+	 * @return array
+	 */
+	public function send_generation_request( $url ) {
+		$response = wp_remote_post(
+			self::API_URL,
+			[
+				// This filter is documented in inc/Engine/CriticalPath/CriticalCSSGeneration.php.
+				'body' => apply_filters(
+					'rocket_cpcss_job_request',
+					[
+						'url' => $url,
+					]
+				),
+			]
+		);
+
+		return $this->prepare_response( $response, $url );
+	}
+
+	/**
+	 * Prepare the response to be returned.
+	 *
+	 * @since 3.6
+	 *
+	 * @param array|WP_Error $response The response or WP_Error on failure.
+	 * @param string         $url Url to be checked.
+	 * @return array|WP_Error
+	 */
+	private function prepare_response( $response, $url ) {
+		$response_data        = $this->get_response_data( $response );
+		$response_status_code = $this->get_response_status( $response );
+		$succeeded            = $this->get_response_success( $response_status_code, $response_data );
+
+		if ( $succeeded ) {
+			return $response_data;
+		}
+
+		$response_code    = $this->get_response_code( $response );
+		$response_message = $this->get_response_message( $response_status_code, $response_data, $url );
+
+		if ( 200 === $response_status_code ) {
+			$response_status_code = 400;
+		}
+
+		return new WP_Error(
+			$response_code,
+			$response_message,
+			[
+				'status' => $response_status_code,
+			]
+		);
+	}
+
+	/**
+	 * Get the status of response.
+	 *
+	 * @param int   $response_code Response code to check success or failure.
+	 * @param array $response_data Array of data returned from request.
+	 * @return bool success or failed.
+	 * @since 3.6
+	 */
+	private function get_response_success( $response_code, $response_data ) {
+		return (
+			200 === $response_code &&
+			! empty( $response_data ) &&
+			(
+				( isset( $response_data->status ) && 200 === $response_data->status ) ||
+				( isset( $response_data->data ) && isset( $response_data->data->id ) )
+			)
+		);
+	}
+
+	/**
+	 * Get response status code/number.
+	 *
+	 * @since 3.6
+	 *
+	 * @param array|WP_Error $response The response or WP_Error on failure.
+	 * @param null|int       $status Status code to overwrite the response status.
+	 * @return int|string status code|number of response.
+	 */
+	private function get_response_status( $response, $status = null ) {
+		if ( is_null( $status ) ) {
+			$status = wp_remote_retrieve_response_code( $response );
+		}
+
+		return (int) $status;
+	}
+
+	/**
+	 * Get response message.
+	 *
+	 * @since 3.6
+	 *
+	 * @param int    $response_status_code Response status code.
+	 * @param array  $response_data Array of data returned from request.
```

Comment:
The response data is shown as an object below. Can it be an array or object? If it's an object, what is it's data type, i.e. instance of ______? Add that to the `@param`.